dashboard/app/models/concerns/user/assigned_courses_and_scripts.rb

```diff
@@ -0,0 +1,245 @@
+  end
+
+  def assigned_script?(script)
+    section_scripts.include?(script) || section_courses.include?(script&.unit_group)
+  end
+
+  # Checks if there are any launched scripts assigned to the user.
+  # @return [Array] of Scripts
+  def visible_assigned_scripts
+    user_scripts.where("assigned_at").
+      map {|user_script| Unit.where(id: user_script.script.id).select(&:launched?)}.
+      flatten
+  end
+
+  # Checks if there are any launched scripts assigned to the user.
+  # @return [Boolean]
+  def any_visible_assigned_scripts?
+    visible_assigned_scripts.any?
+  end
+
+  # Checks if there are any launched scripts or courses assigned to the user.
+  # @return [Boolean]
+  def assigned_course_or_script?
+    assigned_courses.any? || any_visible_assigned_scripts?
+  end
+
+  # Query to get the user_script the user was most recently assigned.
+  def most_recently_assigned_user_script
+    user_scripts.
+      where.not(assigned_at: nil).
+      reorder(assigned_at: :desc).
+      first
+  end
+
+  # Get the UnitGroupUnit for the most recently assigned UserScript.
+  def most_recently_assigned_unit_group_unit
+    unit = most_recently_assigned_user_script&.script
+    return unless unit
+    # UserScript doesn't record the UnitGroup the user was in, so we will assume
+    # it is the most recently created section.
+    section = sections_as_student.select {|s| !s.hidden && s.script_id == unit.id}.last
+    Queries::Courses.unit_group_unit(unit, section&.unit_group)
+  end
+
+  # Get script object of the user_script the user was most recently
+  # assigned.
+  def most_recently_assigned_script
+    most_recently_assigned_user_script&.script
+  end
+
+  def can_access_most_recently_assigned_script?
+    return false unless script = most_recently_assigned_user_script&.script
+
+    !script.pilot? || script.has_pilot_access?(self)
+  end
+
+  # Check if the user's most recently assigned script is associated with at least
+  # 1 live section they are enrolled in.
+  def most_recent_assigned_script_in_live_section?
+    recent_assigned_script_id = most_recently_assigned_script.id
+    sections_as_student.any? {|section| section.script_id == recent_assigned_script_id && section.hidden == false}
+  end
+
+  # Figures out the unique set of scripts assigned to sections that this user
+  # is a part of. Includes default scripts for any assigned courses as well.
+  # @return [Array<Unit>]
+  def section_scripts
+    all_scripts = []
+    all_sections.each do |section|
+      if section.script.present?
+        all_scripts << section.script
+      elsif section.unit_group.present?
+        all_scripts.concat(section.unit_group.default_units)
+      end
+    end
+
+    all_scripts
+  end
+
+  # Query to get the user_script the user made the most recent progress
+  # in.
+  def user_script_with_most_recent_progress
```

Comment:
This method was also refactored to account for the ordering issue outlined above.